Cython/Compiler/ExprNodes.py

```diff
@@ -14272,7 +14272,32 @@ def analyse_type_annotation(self, env, assigned_value=None):
             arg_type = annotation.analyse_as_type(env)
 
         if arg_type is None:
-            warning(annotation.pos, "Unknown type declaration in annotation, ignoring")
+            if annotation.is_name:
+                # Validate annotation in form `var: type`
+                if not env.lookup(annotation.name):
+                    warning(annotation.pos, "Unknown type declaration '%s' in annotation, ignoring" % annotation.name, level=1)
+            elif annotation.is_attribute:
+                # Validate annotation in form `var: module.type`
+                if not env.lookup(annotation.obj.name):
+                    # `module` is undeclared
+                    warning(annotation.pos,
+                            "Unknown type declaration '%s.%s' in annotation, ignoring" % (annotation.obj.name, annotation.attribute),
+                            level=1)
+                elif annotation.obj.is_cython_module:
+                    # `module` is cython
+                    module_scope = annotation.obj.analyse_as_module(env)
+                    if module_scope and not module_scope.lookup_type(annotation.attribute):
+                        warning(annotation.pos,
+                                "Unknown type declaration '%s.%s' in annotation, ignoring" % (annotation.obj.name, annotation.attribute),
+                                level=1)
+                else:
+                    module_scope = annotation.obj.analyse_as_module(env)
+                    if module_scope and module_scope.pxd_file_loaded:
+                        warning(annotation.pos,
+                                "Unknown type declaration '%s.%s' in annotation, ignoring" % (annotation.obj.name, annotation.attribute),
```

Comment:
The other thing that might help - the `AnnotationNode` stores a string representation of the annotation already so you don't necessarily need to construct the string yourself. Not sure whether that lets you simplify too much might it might be useful

Reply:
Fixed.